### DISCUSSION TITLE
rshell: timed out or error in transfer to remote

### DISCUSSION THREAD

**Discussion by smhodge42** [micropython/micropython#16488] [General]
Previous posts indicate this error is not new. However, I have updated the version of rshell to the latest and I am still getting this error. It occurs randomly as nearly as I can tell, approximately half the time, when using the rsync option. It does not seem related to file size. I am wondering if there is a workaround or an rshell option that can be used. This is with an Espressif ESP32-S3-DevKitC, no PSRAM, board that is otherwise working fine once the transfer of files succcesfully occurs. I can always reliably use rshell to open a REPL session. Thanks.

**Comment by davefes** [comment]
Have you been able to do the same on a ESP32 WROOM?  I have had this problem on some STM32 variants.

**Comment by smhodge42** [comment]
I’ve only ever used the board I referred to. It never used to happen until some months ago, and the only that happened over time was that the code grew in size: more files and bigger files. Most of the files are mpy but I cannot say for sure if it only occurs with them.

**Comment by davefes** [comment]
Try uploading a file that is less than 1K. Loading files bigger than 1.5K was the problem I had with the STM32WL55 (Nucelo).  There appears to be some issue with USB comms "getting in the way" with file upload to the flash.  I don't how the USB is setup on your variant.

**Comment by IHOXOHI** [comment]
mpremote works fine,  you can find it in the micropython souce code (in tools/mpremote)..., I don't know with mip files... but maybe the problem comes from them, ..., or the form to work with them... it's a new form to operate, ..., the www is strange too...

**Comment by smhodge42** [comment]
On a whim I moved the USB connection from the “UART” connector on the DevKitC board to the one labeled “USB”.  Issue solved. I have successfully used rshell to sync files well over 50 times since, without a single error. The UART connection continues to work just fine for stdout, aka “print”, use.

I have tried mpremote but insofar as I can tell it does not have the equivalent of rshell’s “rsync” command that automatic uploads all changed files to the device. It seems like it only copies one file at a time w/o checking if they need updating.  If anyone knows of a way to use mpremote to do this, I’d love to hear how.

**Reply by peterhinch** [reply]
> I have tried mpremote but insofar as I can tell it does not have the equivalent of rshell’s “rsync”

mpremote does now do `rsync` when copying: see [this PR](https://github.com/micropython/micropython/pull/11777). Whereas `rshell` uses time to detect difference, `mpremote` uses hashes.

**Reply by IHOXOHI** [reply]
thanks for yours precious precisions...
